### PR TITLE
Remove the transformation untag_int(tag_int x) = x.

### LINKED ISSUE #9028: Miscompilation due to Cmmgen tagging optimisations
`ocamlopt` miscompiles the function `f` below:
```
let f n = ((n lsl 1) + 1) / 2

let r = ref max_int
let () = Printf.printf "%d\n" (f !r)
```
Bytecode prints `0` but native-code prints 4611686018427387903 (i.e. max_int).

The bug is to do with the interaction between `tag_int`, `untag_int` and arithmetic. The code is structured so that `untag_int (tag_int c)` simplifies to `c`, which is valid if `c` is a 63-bit integer sign-extended to 64 bits (that is, its top two bits are equal, as is usually the case for untagged 63-bit integers). This is implemented by pattern-matching in `untag_int`:
```ocaml
let untag_int i dbg =
  match i with
  ...
  | Cop(Caddi, [Cop(Clsl, [c; Cconst_int (1, _)], _); Cconst_int (1, _)], _) ->
      c
  ...
```
This pattern match applies even if `c` is not in sign-extended form. The easiest way to trigger this is to manually construct the offending pattern (see `(n lsl 1) + 1` in the testcase, which simplifies to this), but I suspect there are others.

This means that `untag_int` will sometimes return something which is not in sign-extended form. However, several arithmetic operations (e.g. division) assume that the result of `untag_int` is in sign-extended form and produce the wrong result if it's not.

I believe this is the same issue as #2671. (The fix for that issue fixes it for `Int32.to_int`, but the problem can occur with other functions as well)

(I'm currently working on a patch for this)

### PR DESCRIPTION
**Pull request by stedolan**
This transformation is incorrect for certain large values of x, as it can get the wrong value in the high bit.

This issue was previously noticed for the case of `Int32.to_int` in #2671 (some 18 years ago!). However, the problem can occur in other situations too.

This optimisation no longer seems to be especially important. (I suspect it was when it was first introduced, but other Cmm-level optimisations now cover most of the important cases). In a large JS-internal program (tens of thousands of OCaml files, resulting in a 116MB executable), the optimisation triggers only 377 times.

The most important case where this untagging optimisation applies is when writing a byte or a 16-bit value into a string or bigstring. (In this case, the optimisation is valid, as the high bit is not used). So, this patch adds `untag_int_16` for when the upper bits are not useful, recovering the optimisation in 307 of the 377 occurrences.

Fixes #9028.

### CODE REVIEW

File: asmcomp/cmm_helpers.mli
Comment (lthls) [inline asmcomp/cmm_helpers.mli]: I don't really like the name `untag_int_16`. If I understand correctly, this is the version of `untag_int` that you can call when you know that the argument expression cannot be affected by overflows. A name like `untag_int_assume_no_overflow` would be more precise, tough a bit too verbose.

File: testsuite/tests/regression/pr9028/pr9028.ml
Comment (lthls) [inline testsuite/tests/regression/pr9028/pr9028.ml]: Is there a reason to use references ? My preference would be `Sys.opaque_identity` if the aim is only to prevent constant propagation.

**Review by lthls — APPROVED** [review]
I'd like to see some kind of typing information on cmm (mainly the number of bits that are relevant) that would make it easier to reason on correctness of optimizations.
But this would be a lot of work, and this pull request already improves the situation, so I'm approving it.

**Comment by mshinwell** [comment]
We should probably have a second review on this one, since it is subtle, before merging.  (I would volunteer but I've already been discussing this with @stedolan ; someone independent would be better.)

File: testsuite/tests/regression/pr9028/pr9028.ml
Comment (stedolan) [reply testsuite/tests/regression/pr9028/pr9028.ml]: Only because I was testing this file with (very) old versions of OCaml! I'll change to `Sys.opaque_identity`

File: asmcomp/cmm_helpers.mli
Comment (stedolan) [reply asmcomp/cmm_helpers.mli]: No, that's not the semantics. `untag_int_16 e` can be called for any `e`. It is guaranteed only that the low 16 bits of `untag_int_16 e` match the low 16 bits of `e`, and garbage may be returned in the upper bits. (It's not an assumption about the argument expression, nor about overflows, but a statement that you don't care about some bits of the result)

**Comment by xavierleroy** [comment]
I was expecting your fix to be along the lines of #9006 : make `untag_int` safe, i.e. producing all `word_size` bits correctly, and add and use `untag_int_except_top_bit` (or some other name), producing `word_size - 1` correct low bits and possibly garbage in the top bit.  

Your `untag_int_16` is pretty much the `untag_int_except_top_bit` I'm talking about, but like @lthls I find the name confusing.  Also, I was expecting `untag_int_except_top_bit` to be used in more places where the top bit doesn't matter.  

This said, I can't find any place where `untag_int` is called and the top bit doesn't matter, so maybe all my expectations are wrong.

**Comment by stedolan** [comment]
I have now managed to confuse two people who know this code deeply, so my name is definitely wrong!

`untag_int_except_top_bit` is indeed what I was going for. (I couldn't find anywhere that cared about a bit beyond bit 16 but didn't also care about the top bit, so I picked a name / interface that only specified the low 16 bits)

**Comment by stedolan** [comment]
I've removed `untag_int_16` and instead added `ignore_high_bit_int` (by analogy with `ignore_low_bit_int`). What used to be `untag_int_16 e` is now `ignore_high_bit_int (untag_int e)`, but the generated code's the same.

File: asmcomp/cmmgen.ml
Comment (xavierleroy) [inline asmcomp/cmmgen.ml:495]: The case `_caml_int` could be merged with the cases `_sint8`...`_uint16`.  On the other hand, maybe you prefer to have the same case analysis as for "bigarrayset", lines 503-508.

File: testsuite/tests/regression/pr9028/pr9028.ml
Comment (xavierleroy) [reply testsuite/tests/regression/pr9028/pr9028.ml]: Last chance to use `opaque_identity` instead of references, or leave the test as is.

**Review by xavierleroy — APPROVED** [review]
Looks good to me.  I'd love to see this merged before 4.10 is branched.  One tiny comment and one reminder below, but the code is mergeable as is.

File: asmcomp/cmmgen.ml
Comment (stedolan) [reply asmcomp/cmmgen.ml:495]: Yes, I wanted it to match `bigarrayset`.

File: testsuite/tests/regression/pr9028/pr9028.ml
Comment (stedolan) [reply testsuite/tests/regression/pr9028/pr9028.ml]: Thanks, forgot about this. Fixed now.

**Comment by xavierleroy** [comment]
Excellent!  Let's merge.

**Pull request merged**
